src/Features/Core/Portable/Completion/CompletionService.cs

```diff
@@ -206,44 +207,99 @@ public virtual TextSpan GetDefaultCompletionListSpan(SourceText text, int caretP
             CompletionHelper completionHelper,
-            ImmutableArray<(CompletionItem item, PatternMatch? match)> itemsWithPatternMatch)
+            ImmutableArray<(CompletionItem item, PatternMatch? match)> itemsWithPatternMatch,
+            string filterText)
         {
-            var bestItems = ArrayBuilder<(CompletionItem, PatternMatch?)>.GetInstance();
-            foreach (var pair in itemsWithPatternMatch)
+            // It's very common for people to type expecting completion to fix up their casing,
+            // so if only lowercase characters were typed so far, we'd loosen our standard on comparing items
+            // in case-sensitive manner and take into consideration the MatchPriority as well.
+            // i.e. when everything else is equal, then if item1 is a better case-sensitive match but item2 has higher 
+            // MatchPriority, we consider them equally good match, so the controller will later have a chance to
+            // decide which is the best one to select.
+            var isFilterTextLowercaseOnly = !filterText.Any(char.IsUpper);
+
+            // Keep track the highest MatchPriority of all items in the best list.
+            var highestMatchPriorityInBest = int.MinValue;
+            using var _1 = ArrayBuilder<(CompletionItem, PatternMatch?)>.GetInstance(out var bestItems);
+
+            // This contains a list of items of worse case-sensitive match but higer MatchPriority than all best ones.
+            using var _2 = ArrayBuilder<(CompletionItem, PatternMatch?)>.GetInstance(out var similarItemsWithHigerMatchPriority);
+
+            foreach (var (item, match) in itemsWithPatternMatch)
             {
                 if (bestItems.Count == 0)
                 {
                     // We've found no good items yet.  So this is the best item currently.
-                    bestItems.Add(pair);
+                    bestItems.Add((item, match));
+                    highestMatchPriorityInBest = item.Rules.MatchPriority;
+                    continue;
                 }
-                else
+
+                var (bestItem, bestItemMatch) = bestItems.First();
+                var comparison = completionHelper.CompareItems(
+                    item, match, bestItem, bestItemMatch, out var onlyDifferInCaseSensitivity);
+
+                if (comparison == 0)
                 {
-                    var (bestItem, bestItemMatch) = bestItems.First();
-                    var comparison = completionHelper.CompareItems(pair.item, pair.match, bestItem, bestItemMatch);
-                    if (comparison < 0)
+                    // This item is as good as the items we've been collecting.  We'll return 
+                    // it and let the controller decide what to do.  (For example, it will
+                    // pick the one that has the best MRU index).
+                    bestItems.Add((item, match));
+
+                    if (highestMatchPriorityInBest < item.Rules.MatchPriority)
```

Comment:
[comparison of two items](https://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/Completion/CompletionHelper.cs,244054fec2305c68) is based on criteria like text-based matching, whether it's expanded item,  whether it's a keyword and whether it should be pre-selected. Interestingly, pre-selection is specified via MatchPriority, but it's just one specific value, so the MatchPriority isn't used as part of comparison in general (i.e. if one item has higher MatchPriority than the other, but not [Preselect](https://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/Completion/MatchPriority.cs,828604b92747510a,references), then it's not taken into consideration)

The essence of this change is just that we don't out right disregard the items that don't match the casing but have higher priority than the ones do, i.e. (in pseudo code) if `Compare(item1, item2) < 0` && `onlyDifferInCaseSensitivity == true` && `item2.MatchPriority > item1.MatchPriority` && `isFilterTextLowercaseOnly`, then we'd want to consider item2 to be a better candidate for selection.

> what's an example of a case where comparison == 0 but MatchPriority differs? 

One that I can think of is `[nuint](https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/Completion/KeywordRecommenders/NuintKeywordRecommender.cs,21)` vs `null`, when user typed "n". However, please note that this is NOT the issue this change is trying to resolve, when comparison is 0, then the select IS determined by MatchPriority.

The scenario that prompt the fix is selecting between the conversion symbol `(short)` and a method symbol `Should`, when lowercase "s" is typed. We'd like to de-prioritize conversions since it's relatively rare people need them, but with the way comparison works today, case-sensitive match is always preferred over MatchPriority.

Reply:
Makes sense, appreciate the detailed explanation!